Adding Narrator texts for Preview changes pane

dominoFire opened the pull request:
## Bug

Fixes: 
- https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1049251
- https://github.com/nuget/home/issues/9077

Regression: No
* Last working version:   
* How are we preventing it in future:   

## Fix

Details: Add a localized name for the pane, so that, the narrator can read with more context about the pane.

## Testing/Validation

Tests Added: Yes. Unit tests for new fields
Validation:  See screenshots below

### Before

Pane has no text for narrator

![image](https://user-images.githubusercontent.com/1192347/73410125-2d513800-42b6-11ea-843d-3964d2ac3539.png)


### After

- Pane has text 'Package changes'. The pane is read as 'Preview changes pane'
- The packages listed in the pane are read as '<packageId> version <packageVersion>'. In the example below, the narrator will read 'WindowsAzure.Storage version 9.3.1 to WindowsAzure.Storage version 9.3.3'

![image](https://user-images.githubusercontent.com/1192347/73579614-bfcd1500-4437-11ea-8cf6-7a67acbaa482.png)

Commit:
Adding Narrator text for Preview changes pane

Commit:
Removed hardcoded strings. Added accesible name for package identities in Preview window

Commit:
Added tests for PackageIdentityResult

Commit:
Polish

dtivel commented on test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/Actions/PackageIdentityResultTests.cs:
```diff
@@ -0,0 +1,21 @@
+    public class PackageIdentityResultTests
+    {
+        [Fact]
+        public void AccessibleName_ReturnsMessage_CultureNeutral()
```
`AutomationName_WhenCultureIsNeutral_ReturnsMessage`

dominoFire replied:
Renamed.

dominoFire replied:
Renamed.

dtivel commented on test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/Actions/PackageIdentityResultTests.cs:
```diff
@@ -0,0 +1,21 @@
+        [Fact]
+        public void AccessibleName_ReturnsMessage_CultureNeutral()
+        {
+            PackageIdentityResult result = new PackageIdentityResult("test.package", new NuGetVersion(0, 0, 1));
```
`var`

dominoFire replied:
Changed.

dominoFire replied:
Fixed.

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/PackageIdentityResult.cs:
```diff
@@ -0,0 +1,31 @@
+    /// <summary>
+    /// Represents a package identity with an acessible name in the Preview Window
+    /// </summary>
+    public class PackageIdentityResult : PackageIdentity
```
Not crazy about this naming convention.  What do you think about `AccessiblePackageIdentity`?

Make it `sealed`.

dominoFire replied:
I was trying to follow the pattern name in `PackageIdentityResult` and [`UpdatePreviewResult`](https://github.com/NuGet/NuGet.Client/blob/319fa107e1c90be67da6e1f57f80de9398dfdc3c/src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UpdatePreviewResult.cs#L8). The latter also creates an acessible name for a package update in the Preview Windows in the UI. I will change `PackageIdentityResult` to `sealed` and rename it.

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/PackageIdentityResult.cs:
```diff
@@ -0,0 +1,31 @@
+        {
+        }
+
+        public string AutomationName => string.Format(
```
This results in several string allocations each time the property is called.  If the value won't change, can't we just create the value once?

dominoFire replied:
It seems that those names are created only once in this [method in UIActionEngine]( https://github.com/NuGet/NuGet.Client/blob/319fa107e1c90be67da6e1f57f80de9398dfdc3c/src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UIActionEngine.cs#L840).

Later, the [results](https://github.com/NuGet/NuGet.Client/blob/319fa107e1c90be67da6e1f57f80de9398dfdc3c/src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UIActionEngine.cs#L374) are only read through all the code paths.

I will change it to a Lazy value.

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UpdatePreviewResult.cs:
```diff
@@ -21,6 +22,10 @@ public override string ToString()
         }
 
-        public string AutomationName => $"{Old} to {New}";
+        public string AutomationName => string.Format(
```
See my other comment.  Can't we create this value once and only once?

dominoFire replied:
Sure.

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Resources.resx:
```diff
@@ -874,4 +874,18 @@ Please see https://aka.ms/troubleshoot_nuget_cache for more help.</value>
+  </data>
+  <data name="Preview_PackageUpdate" xml:space="preserve">
+    <value>{0} version {1} to {2} version {3}</value>
+    <comment>{0} = Old package Id, {1} = Old package version, {2} = New package version, {2} = New package version</comment>
```
Shouldn't it be:
```
{0} = Old package ID, {1} = Old package version, {2} = New package ID, {3} = New package version
```

dominoFire replied:
Fixed.

dominoFire replied:
Fixed.

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Resources.resx:
```diff
@@ -874,4 +874,18 @@ Please see https://aka.ms/troubleshoot_nuget_cache for more help.</value>
+    <comment>{0} = Old package Id, {1} = Old package version, {2} = New package version, {2} = New package version</comment>
+  </data>
+  <data name="Preview_UnknownProject" xml:space="preserve">
+    <value>Unknown Project</value>
```
Does "project" need to be capitalized?

dominoFire replied:
The [original string](https://github.com/NuGet/NuGet.Client/blob/a746cd6a01dfd22187375ee9c5a0d9b98676020a/src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/PreviewResult.cs#L35) appears capitalized.  That string is for the corner case where PM UI doesn't know the project name in the Preview Window Dialog. A sample text would look like the following:

```
Unknown Project

Installing:

Microsoft.Azure.KeyVault.Core.1.0.0
WindowsAzure.Storage.9.3.3
```

vs.


```
Unknown project

Installing:

Microsoft.Azure.KeyVault.Core.1.0.0
WindowsAzure.Storage.9.3.3
```

dominoFire replied:
I will make it lower case.

dtivel commented on test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/Actions/PackageIdentityResultTests.cs:
```diff
@@ -0,0 +1,21 @@
+
+            var msg = result.AutomationName;
+
+            Assert.Equal("test.package version 0.0.1", msg);
```
`msg` seems like its only purpose is to preserve the result of a property which may have different values over time.  Is that true?  If no, just replace `msg` with `result.AutomationName` and delete the local `msg`.  If yes, change `var msg` to `string actualResult`.

dominoFire replied:
`result.AutomationName` doesn't change over time (see comments on the property). Thus, I will just use property in the assertion.

Commit:
Addressed feedback

donnie-msft commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PreviewWindow.xaml:
```diff
@@ -64,7 +64,8 @@
       <ScrollViewer
         HorizontalScrollBarVisibility="Auto"
-        IsTabStop="True">
+        IsTabStop="True"
```
Was this still needed?

dominoFire replied:
Yes. The line below makes the narrator to read a more meaningful name for the Preview Changes Pane

donnie-msft commented on test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/Actions/AccessiblePackageIdentityTests.cs:
```diff
@@ -0,0 +1,28 @@
+        {
+            var result = new AccessiblePackageIdentity("test.package", new NuGetVersion(0, 0, 1));
+
+            Assert.Equal("test.package version 0.0.1", result.AutomationName);
```
Probably fine... but maybe `Version.ToNormalizedString()` would be better in case we change how that looks at some point?

dominoFire replied:
Good thought. If the implementation of `Version.ToNormalizedString()` changes, then, the test should fail to indicate that the accessible name needs to be addressed. I'm thinking of making 2 assertions, one for the hardcoded string (actual) and one for comparing against `Version.ToNormalizedString()`

kartheekp-ms replied:
I think assert with hard-coded string is unnecessary because the actual value doesn't change in both cases and the underlying string `AutomationName`  is also constructed by invoking `Version.ToNormalizedString()`.

donnie-msft commented on test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/Actions/AccessiblePackageIdentityTests.cs:
```diff
@@ -0,0 +1,28 @@
+        {
+            var result = new AccessiblePackageIdentity(new PackageIdentity("test.package", new NuGetVersion(0, 0, 1)));
+
+            Assert.Equal("test.package version 0.0.1", result.AutomationName);
```
Same comment as other test...

dominoFire replied:
See [reply](https://github.com/NuGet/NuGet.Client/pull/3217#discussion_r375614674).

kartheekp-ms replied:
I think assert with hard-coded string is unnecessary because the actual value doesn't change in both cases and the underlying string `AutomationName`  is also constructed by invoking `Version.ToNormalizedString()`.

donnie-msft reviewed (commented):
Just some observations...

Commit:
Double checking on tests

Commit:
Reverted accidental change on NuGet.VisualStudio.Client.csproj

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Resources.resx:
```diff
@@ -874,4 +874,18 @@ Please see https://aka.ms/troubleshoot_nuget_cache for more help.</value>
+  </data>
+  <data name="Preview_PackageUpdate" xml:space="preserve">
+    <value>{0} version {1} to {2} version {3}</value>
+    <comment>{0} = Old package ID, {1} = Old package version, {2} = New package ID, {2} = New package version</comment>
```
Last {2} is {3}.

dominoFire replied:
Fixed 😁

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Resources.resx:
```diff
@@ -874,4 +874,18 @@ Please see https://aka.ms/troubleshoot_nuget_cache for more help.</value>
+  <data name="Accessibility_PackageChanges" xml:space="preserve">
+    <value>Package changes</value>
+  </data>
+  <data name="Accesibility_PackageIdentity" xml:space="preserve">
```
Accesibility -> Accessibility

dominoFire replied:
Fixed.

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UpdatePreviewResult.cs:
```diff
@@ -9,18 +11,24 @@ public class UpdatePreviewResult
     {
         public PackageIdentity Old { get; }
         public PackageIdentity New { get; }
+        private Lazy<string> _automationNameLazy;
```
`readonly`

I'm not sure you need `Lazy<T>`.  The fact that you're creating an `UpdatePreviewResult` means you'll be accessing the value anyway, right?  Why not just make this a `readonly string`?

dominoFire replied:
After an `UpdatePreviewResult` object is created, the `AutomationName` property is used in the Preview Dialog UI. There's no benefit of creating the string in a delayed manner. Since this property is never modified, it makes sense to use a read only member. I changed to a read only string property named `AutomationName`

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/AccessiblePackageIdentity.cs:
```diff
@@ -0,0 +1,35 @@
+    /// </summary>
+    public sealed class AccessiblePackageIdentity : PackageIdentity
+    {
+        private Lazy<string> _automationNameLazy;
```
`readonly`

Is `Lazy<T>` really important?  Does it buy us much?  Can it just be `readonly string`?

dominoFire replied:
Same as https://github.com/NuGet/NuGet.Client/pull/3217#discussion_r376162554

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/AccessiblePackageIdentity.cs:
```diff
@@ -0,0 +1,35 @@
+    {
+        private Lazy<string> _automationNameLazy;
+
+        public AccessiblePackageIdentity(string id, NuGetVersion version)
```
Maybe I missed it, but [I only see the other constructor `AccessiblePackageIdentity(PackageIdentity)` being called](https://github.com/NuGet/NuGet.Client/pull/3217/files#diff-cf33c7d7a6896e312b8b86f1e05f557fR909-R914).

This constructor looks to be unnecessary.  If it is, remove it and move `_automationNameLazy` initialization into the other constructor.

dominoFire replied:
I implemented that constructor to avoid making lot of changes in which the original `PackageIdentity` constructor is used.

dominoFire replied:
Discussed offline. I realized that the changes are only on tests. So, I only need one constructor in production source code. I will remove the redundant constructor.

Commit:
Addredded feeback 2

- Removed redundant constructor
- Added test
- Fixed typos
- Moved Lazy to readonly strings

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/AccessiblePackageIdentity.cs:
```diff
@@ -0,0 +1,29 @@
+    public sealed class AccessiblePackageIdentity : PackageIdentity
+    {
+        private readonly string _automationName;
+        public string AutomationName => _automationName;
```
```suggestion
        public string AutomationName { get; }
```

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UpdatePreviewResult.cs:
```diff
@@ -9,18 +11,24 @@ public class UpdatePreviewResult
     {
         public PackageIdentity Old { get; }
         public PackageIdentity New { get; }
+        private readonly string _automationName;
```
Why? 
That's the benefit? 
Just use a readonly property.

dominoFire replied:
I wasn't aware of c# read only properties. Thanks!

nkolev92 commented on test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/Actions/AccessiblePackageIdentityTests.cs:
```diff
@@ -0,0 +1,22 @@
+            var version = new NuGetVersion(0, 0, 1);
+            var result = new AccessiblePackageIdentity(new PackageIdentity("test.package", version));
+
+            Assert.Equal("test.package version 0.0.1", result.AutomationName);
```
Why do you have 2 very similar asserts.

dominoFire replied:
One is for detecting the most simple case. The second is to detect , in a case of a test failure, where the possible error is.

nkolev92 replied:
I don't think I follow. 

You would get the same error message in both cases.

nkolev92 commented on test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/Actions/UpdatePreviewResultTests.cs:
```diff
@@ -0,0 +1,27 @@
+                new PackageIdentity("updated.package", oldVersion),
+                new PackageIdentity("updated.package", newVersion));
+
+            Assert.Equal("updated.package version 0.0.1 to updated.package version 0.0.2",
```
Same here, the assert is very similar.

Commit:
Addressed feedback on strings
